### ISSUE TITLE
instability of a QG jet

### ISSUE THREAD

**Issue by francispoulin** [pyqg/pyqg#71]
Inspired by the recent advances in pyqg, I would like to put together another QG example.  Basically, I want to reproduce some of the results of Glenn's from 1987.

http://journals.ametsoc.org/doi/abs/10.1175/1520-0485(1987)017%3C1408:NWACVS%3E2.0.CO;2

He uses the reduced gravity BT QG model but one issue is that the streamfunction is not periodic in one direction.  One way to do this is to decompose the model into a basic state and perturbation

psi = Psi + psi'
q    = Q   + q'
u    = U   + u'

The fully nonlinear governing equations are then

partial_t q' + J(Psi, q') + J(psi', Q) + J(psi',q') = 0

and we can look for solutions that are doubly periodic.  

I know how to solve this in general but not sure how to fit this into the context of pyqg.  I looked in BT_model and the evolution equations are not there.

Ryan: any ideas how best to do this? 

Another option is to look how to use DCT in one direction.  I thought the first option would be easier but am very interested to know what you all think. 

If you think this is too much of a pain for the moment I don't have to do it now but I thought I'd mention it since it would be a nice example. Doing the linear stability would be nice and easy too.

**Comment by rabernat** [comment]
This is a great suggestion. It seems like there are two different things being suggested here

1. Allowing for spatially variable background states
2. Enforcing different lateral boundary conditions

I think both should eventually be implemented. In particularly, 1 would allow us to include topography through the pv gradients, something I would be very interested in. 2 would open a whole new range of scenarios, including gyres and stuff.

It would be quite easy to implement 1. The model already adds a constant background U and Qy. The kernel would just have to be modified to allow these to be arrays instead of constants. An important part of this would be to keep things backwards compatible with the examples we already have. All of the action is in kernel.pyx.

As for 2, pyfftw simply does not support DCT. However, the fftw package itself does.
http://www.fftw.org/doc/Real-even_002fodd-DFTs-_0028cosine_002fsine-transforms_0029.html
In order to take advantage of this, we would need to bypass pyfftw and make our own cython wrapper for fftw. I'm sure it would be tricky, but there are plenty of examples to go on, e.g. [daedalus](https://bitbucket.org/dedalus-project/dedalus/src/150221b94a6a75712cb12098c2585287ffa8bffb/dedalus/libraries/fftw/fftw_wrappers.pyx?at=default). This would have the added benefit of removing the pyfftw dependency, which can be a pain for some people.

Numpy does not support DCT either, but scipy.fftpack does
http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.fftpack.dct.html

So in summary, yes, 1 is easier, and I fully encourage you to implement it. If you go this route, we will want to discuss the desired API for setting the background state as a group. For example, do you specify Psi? Q? Both? Do we check that they satisfy the inversion relation and give no tendency, i.e. J(Psi, Q)==0?

**Comment by rabernat** [comment]
> I looked in BT_model and the evolution equations are not there.

What do you mean by this?

**Comment by francispoulin** [comment]
Thanks for your your thoughts Ryan.

When I said that I didn't see the evolution equations I meant I didn't see where J(psi,q) was calculated and then advanced using the AB3 method.  I think I see that it is in model.py.  Because pyqg solves everything of the same form, this is always the same I presume.  Probably don't want to change that unless we have it. 

I agree that changing the velocities and other fields to be spatial variable should be easy. Not sure where it is explicitly specified that U is constant, maybe it isn't even declared.  Not sure but should better understand this.

Thanks for the info on other transform methods.  There is another way to avoid this, but is not quite as efficient.  Also, I should have said DST but that's a minor detail.  If we want to compute the derivative of a field that satisfies zero Dirichlet boundary conditions at y=0,L for example, we can do this by extending the field. If we extend the field about y=L in an odd way (flip it but change the sign), then the new field is periodic.  I have done this in a 1-layer QG code in python and I think it works nicely.  The differences are,

a. Define the wavenumber based on the length being bigger and the number of grid points being double.
b. Must extend the field before we compute derivatives.
c. Must then restrict our attention to the physical part of the domain.  

As I said I have sample code that does this but not sure how best to show that on github.

Ryan: I imagine these are easy enough to do?

This isn't necessarily as efficient as other methods and it does require more memory but the good news is that it does work and is easy to do.  

In terms of a strategy:

1. I can try creating a new example that specifies U to be something that is spatially variable, but periodic for starters.  

2. Maybe I should think about how we could have a switch between doubly periodic and a zonal channel.

This won't happen in the next couple of days but I will think about it when I have time.

**Comment by rabernat** [comment]
> Not sure where it is explicitly specified that U is constant, maybe it isn't even declared.

Everything happens in the kernel's advection routine
https://github.com/pyqg/pyqg/blob/master/pyqg/kernel.pyx#L341
(Note that this is cython.)

**Comment by rabernat** [comment]
The domain-extending method sounds very expensive compared to DST / DCT. I brought this up over at pyfftw (hgomersall/pyFFTW#39), and it sounds like they might consider adding it, which would save us a lot of pain.

**Comment by rabernat** [comment]
> I can try creating a new example that specifies U to be something that is spatially variable, but periodic for starters.

Yes, this is a good way to proceed. Parallel shear flow is the next step up in complexity. After that comes a non-zonally-symmetric background flow, which requires a self-consistent background PV gradient. That is the part I am worried about--how to we enforce J(Psi, Q) = 0?

**Comment by MFJansen** [comment]
In a (fully nonlinear) doubly periodic QG model, the difference between "background flow" and "perturbation" should become relevant only through the boundary condition (which is periodic in the perturbation).  If the background shear / interface height is a function of y that would modify what's a "perturbation" versus "background", but the full PV/flow field should dependent only on the meridional integral of the background shear (i.e. the interface height jump across the interface). With that in mind, I'm not sure if it is particularly useful to include a non-constant background flow. Topography, on the other hand, would be useful, and should be straightforward to implement. I'm hoping that I might find some time to do something on pyQG next weekend.

**Comment by rabernat** [comment]
Good point @MFJansen.

We should think about what our top priorities for our next release are. I have been trying to file issues for each of them with the 0.2 milestone tag:
https://github.com/pyqg/pyqg/milestones/0.2

**Comment by francispoulin** [comment]
Thanks @MFJansen for your thoughts.  

Whether it would be useful to include background stratification depends on our aims. I for one would like to see pyqg look at the Barotropic instability of a jet, like Glenn did in the paper I mentioned.  One option is to put the jet in the basic state, but then the equations change significantly.  A second, and perhaps easier option, is to allow for non-periodic BCs.  I agree that one needs to be careful about how the meridional integral of the background shear evolves, and Glenn has shown us how that is done.  But that involves solving a PDE with 1D.  Relatively easy to do in principle.  

It would be great if DCT/DST could be included in pyfftw but there is no telling if/when that will happen.  I don't know how to do a wrapper for FFTW so I'm a bit lost there.  I do know how to do these extensions, but if that is undesirable, I guess we can wait to see what happens with pyfftw.

**Comment by MFJansen** [comment]
Are you interested in solving the linearized problem? If so, I agree that it would be useful to allow for a meridionally varying background flow  (once we are linearizing it of course matters what exactly we are linearizing around). We of course also would need an option to turn off the non-linear term (or do we have that already? - I admit I'm not up-to-date). I agree that this would be an interesting addition.

**Comment by francispoulin** [comment]
I have python code that solve the linearized problem already.  I would like to include this into pyqg, just not sure how to do it.  Being able to see the result from linear theory, and then looking at the nonlinear evolution would be quite nice I think.  This could generalize nicely to multiple layers but it's the one-layer that is the hard part.    But it would be good to decide what direction we feel is best for pyqg, and the users of course.

**Comment by rabernat** [comment]
There is no option to turn of the background state.

By "linearized problem" are you referring to something different from the linear stability analysis proposed in #37?

Near the top of my list of priorities is the ability to perform automatic linear stability analysis on the existing models, without the additional complication of meridionally varying background flow.

With a meridionally varying background flow, the eigenmodes have both a vertical and meridional structure (rather than normal modes in y). This seems significantly more difficult.

> But it would be good to decide what direction we feel is best for pyqg

The best direction is the one that is most useful for our research and teaching. So @francispoulin if you want these features, we should definitely add them! The challenge is to add them in the most general and abstract way, rather than piecing together a jumble of independent scripts

That's why I think it is useful to think about the API design. How should it work? For example

```python
m = pyqg.BTModel(beta=0)
m.set_background_state(u=np.sin(m.y))
eigenvals, eigenvects = m.linear_stability()
m.run()
```

Or maybe the result of ```m.linear_stability()``` should be a new type of object wraps the results of the stability analysis in an intelligent way.

Thinking in terms of API design can help us clarify how to move forward.

**Comment by rabernat** [comment]
Just posting @francispoulin's example notebook so the discussion can continue here:
https://gist.github.com/rabernat/1aa171fef7c5de2901d3

**Comment by rabernat** [comment]
@francispoulin I like your notebook. It is certainly useful to be able to do those sort of things. The application I am thinking of is when I teach about barotropic instability and the Rayleigh-Fjortoft criteria. We are limited to the few simple examples U(y) profiles that we know how to solve analytically. It would be great to be able to do the linear stability analysis numerically for arbitrary U(y) profiles (and then do the full nonlinear simulation of course).

But your notebook does not answer the question I posed above.

- How do you imagine interacting with this feature through pyqg's [api](https://en.wikipedia.org/wiki/Application_programming_interface)?
- How does generalize to the other models besides barotropic? Can I do the same thing in two-layer QG? SQG?
- How does it connect with the fully nonlinear problem? Can we implement the linear problem without solving the issue of a meridionally varying background state?

**Comment by francispoulin** [comment]
Sorry for not answering your questions @rabernat .  I'll try and answer them here.

Q1) I think the form you suggested above is very intuitive and easy to do. Most of this code could be hidden away in a function.  There would need to be a few more parameters like the wavenumber range to look at, but that would be easy enough to do, I think.  

Q2) Because all of these models have the same form, it does generalize quite well.  For the 2-layer case you get 2 equations and in general for n-layers, n-equations.  The SQG case should be almost identical. I would have to work a little to actually do this but this is the easy part.

Q3) This particular basic state has a non-zero meridional mean and the streamfunction would not be continuous to the north and south.  Something else needs to be done for the nonlinear problem.  The two methods mentioned are 

M1) write a wrapper for fftw that uses DST and 
M2) do periodic extensions. 

I agree that the former is more attractive in terms of efficiency but since other people would need to do this, I don't know when/if that would happen.  I know how to do M2) pretty easily but would have to learn a lot to do M1.

**Comment by rabernat** [comment]
Regarding DST, it looks like my comments over at pyfftw have created a stir (hgomersall/pyFFTW#66). Sounds like they will be releasing a new version that supports DST / DCT soon.

I like your replies, and am getting convinced that this is a good way to go. Two more questions.

- How would the linear stability analysis proposed here relate to that proposed in #37? What happens if you have a constant flow in each layer? (Obviously nothing in the barotropic case, but plenty of interesting stuff in the multi-layer case.) It would be nice to have it be as general as possible.
- Can you easily add friction?

**Comment by francispoulin** [comment]
Thanks for the update @rabernat .  That is encouraging.  At some point if we figure out how to do FFTW wrappers then we might be able to do this ourselves, but I'm a far way from knowing how that works.

Some answer:

Q1) I look at the other QG code.  That seems to assume the velocity is constant and the horizontal is doubly periodic.  That's why the derivatives simply become i*k's.  For the 1-layer problem, we have a scalar equation to solve instead of a N by N system.  For the n-layer problem they have an n by n system, whereas we would have an nN by nN.  I don't think we would want to take n to be too large but in theory it can be done. 

Q2) There is no difficulty in adding friction, but we do need that the steady solution is an exact solution to the nonlinear system of equations.  If you have the basic state that I mentioned, you would need to add some forcing to balance the friction, which is easy enough to do.  So yes is the short answer.

**Comment by rabernat** [comment]
Great. So I think the next step is for you to put together a pull request demonstrating how this would work.

Don't forget to include tests! Lots and lots of tests! The tests are what make this whole thing possible. Don't only test the high-level results. Test every aspect of your code.

**Comment by francispoulin** [comment]
Thanks @rabernat for the advice. I do not have a working model but I did start to put a skeleton together (using gist.github).  

https://gist.github.com/francispoulin/89faf215b870f6b733b1

Unfortunately, I don't understand the framework of pyqg as much as everyone else so I was hoping for some advice, in terms of the set up.  If you look in the places with FJP you will see some questions in terms of arguments.  Any advice would be welcome.

If I can get this properly set up then it *should* be a simple matter of calling the class.  A job best saved for tomorrow.

**Comment by rabernat** [comment]
Why don't you try adding the linear stability method to the main Model class (model.py)?

**Comment by crocha700** [comment]
> I asked a question a while ago about including the linear stability analysis of the Bickley let. I would like to include that and even do a multi-layer version. 

One way to go is to introduced a method into ```model.py``` generalized linear stability analysis in horizontal and vertical space (My next PR will be on a liner stability analysis assuming horizontally constant  base state). May be you could take it from there?

>I also would like to do nonlinear simulations with pyqg. Since the Bickley jet does not have periodic BCs, I think I would need to decompose the flow into a basic state and a perturbation and solve the nonlinear perturbation equations. I know the equations, I'm just not sure how this will fit in the pyqg framework.

We already calculate the linear part of the advection in a [pseudo-spectral sense](http://git.io/vl4On). That would simply have to change to include the horizontal dependence. Also, with Ubg function of (x, y), the base-state PV gradients would also be a function of (x,y). So, [Qx and Qy](http://git.io/vl4ZN) would also have to change. 

I believe the best way to discuss this is to branch off of master and try to implement something, and then share with us via gist, PRs, etc.

**Comment by rabernat** [comment]
> I believe the best way to discuss this is to branch off of master and try to implement something, and then share with us via gist, PRs, etc.

:+1:

**Comment by francispoulin** [comment]
Thanks @crocha700  for the response.  I am happy to push this forward, since you've done so much already.  

I did have a notebook that did the one-layer case.  I would very much like to do the n-layer case (where I realize we won't use n too large).  One question is do you think I should try and implement the the one-layer version or wait until I have the n-layer typed up.  

After I see how you wrote down the equations, one of the docs I'm sure is there, it should be easy enough to do.  It's how to include it that seems mysterious but after I look over the new code maybe it will become apparent.

**Comment by crocha700** [comment]
>One question is do you think I should try and implement the the one-layer version or wait until I have the n-layer typed up.

@francispoulin francispoulin  It is really up to you whether to implement the barotropic version first or not. There are advantages and disadvantages in both choices (I generally try to think more general). It really does not matter. Just brach off of master and start coding it up in ```model.py```, etc.


> After I see how you wrote down the equations, one of the docs I'm sure is there, it should be easy enough to do. It's how to include it that seems mysterious but after I look over the new code maybe it will become apparent.

The nice thing of working with version control and branches is that we can make changes without fearing to break the code. The best way to learn more about the code is to changing it... If something goes wrong, you can always revert changes or abandon that branch altogether...

**Comment by francispoulin** [comment]
I am *trying* to post a PDF, here, not sure if it will work.  It's not shocking but I did write up the equations I want to solve for the n-layer linear stability problem.  It looks pretty simple and after I figure out that @crocha700 did, I hope that I can make it work.

In reading through the existing notes I found a typo in a subscript on Q and am not sure of topography is being double counted.  I will revisit this with a fresh set of eyes and at some point push a fix.

[notation_layered_stability.pdf](https://github.com/pyqg/pyqg/files/26739/notation_layered_stability.pdf)
